tests/std/tests/P0323R12_expected/test.cpp

```diff
@@ -0,0 +1,1865 @@
+            using Input    = std::unexpected<convertible>;
+            using Expected = expected<int, payload_constructors>;
+
+            const Input const_input{in_place};
+            const Expected copy_constructed{const_input};
+            assert(!copy_constructed);
+            assert(copy_constructed.error() == 3);
+            static_assert(noexcept(Expected{const_input}) == should_be_noexcept);
+
+            const Expected move_constructed{Input{in_place}};
+            assert(!move_constructed);
+            assert(move_constructed.error() == 42);
+            static_assert(noexcept(Expected{Input{in_place}}) == should_be_noexcept);
+        }
+
+        { // in place payload
+            using Expected = expected<payload_constructors, int>;
+            const Expected default_constructed{in_place};
+            assert(default_constructed);
+            assert(default_constructed.value() == 0);
+            static_assert(noexcept(Expected{in_place}));
+
+            const Expected value_constructed{in_place, convertible{}};
+            assert(value_constructed);
+            assert(value_constructed.value() == 42);
+            static_assert(noexcept(Expected{in_place, convertible{}}) == should_be_noexcept);
+
+            const Expected ilist_value_constructed{in_place, {1}, convertible{}};
+            assert(ilist_value_constructed);
+            assert(ilist_value_constructed.value() == 1337);
+            static_assert(noexcept(Expected{in_place, {1}, convertible{}}) == should_be_noexcept);
+        }
+
+        { // in place error
+            using Expected = expected<int, payload_constructors>;
+            const Expected default_constructed{unexpect};
+            assert(!default_constructed);
+            assert(default_constructed.error() == 0);
+            static_assert(noexcept(Expected{unexpect}));
+
+            const Expected value_constructed{unexpect, convertible{}};
+            assert(!value_constructed);
+            assert(value_constructed.error() == 42);
+            static_assert(noexcept(Expected{unexpect, convertible{}}) == should_be_noexcept);
+
+            const Expected ilist_value_constructed{unexpect, {1}, convertible{}};
+            assert(!ilist_value_constructed);
+            assert(ilist_value_constructed.error() == 1337);
+            static_assert(noexcept(Expected{unexpect, {1}, convertible{}}) == should_be_noexcept);
+        }
+    }
+
+    constexpr void test_constructors() noexcept {
+        test_constructors<IsNothrowConstructible::Yes, IsExplicitConstructible::Yes>();
+        test_constructors<IsNothrowConstructible::No, IsExplicitConstructible::Yes>();
+        test_constructors<IsNothrowConstructible::Yes, IsExplicitConstructible::No>();
+        test_constructors<IsNothrowConstructible::No, IsExplicitConstructible::No>();
+    }
+
+    template <IsNothrowCopyConstructible nothrowCopyConstructible, IsNothrowMoveConstructible nothrowMoveConstructible,

```

Comment:
I think this is simpler with boolean NTTPs and normal enums, instead of enum-classes and enum-class NTTPs

If you think this approach is better, feel free to keep it though.

Reply:
it would also make it easier to see that all combinations are covered when instantiating and calling the tests.

Reply:
I really do not like boolean parameters

And I would argue that it does not really become more readable. It is ugly no matter what

Reply:
Yeah, there's precedent for the enum class thing, and it does prevent you from using the wrong enum for the wrong parameter, but the code _is_ more readable with normal bools and normal enums (there's a good reason why C enums dump identifiers into the parent scope, I really don't like enum classes at all unless they are used with "using enum").

As with most purely style things I can deal with keeping this as-is, esp if other maintainers agree with you.

Reply:
I agree with @miscco - the safety of avoiding unintentional parameter mixing outweighs the verbosity here.